icepyx/core/visualization.py

```diff
@@ -0,0 +1,312 @@
+            icesat2_files = region.avail_granules(ids=True, cycles=True, tracks=True)[0]
+
+            if not icesat2_files:
+                continue
+
+            is2_bbox_list.append(bbox_i)
+            is2_file_list.append(icesat2_files)
+
+        filelist_tuple = zip(is2_bbox_list, is2_file_list)
+
+        return filelist_tuple
+
+    def generate_OA_parameters(self):
+        """
+        Get metadata from file lists in each 5*5 bbox
+        :return: a list of parameters for OpenAltimetry API query, including
+                 RGT, cycle number, datetime, bounding box, product name
+        """
+        # list of parameters for API query
+        paras_list = []
+
+        # t uple of non-empty bbox and ICESat-2 datafiles
+        filelist_tuple = self.query_icesat2_filelist()
+
+        for bbox_i, filelist in filelist_tuple:
+
+            for fname in filelist:
+
+                if self.product in ['ATL07', 'ATL10']:
+                    temp = intake.source.utils.reverse_format(
+                        format_string="{product:5d}-{HH:02d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{"
+                                      "region:02d}_{release:03d}_{version:02d}.h5",
+                        resolved_string=fname,
+                    )
+
+                else:
+                    temp = intake.source.utils.reverse_format(
+                        format_string="{product:5d}_{datetime:%Y%m%d%H%M%S}_{rgt:04d}{cycle:02d}{region:02d}_{"
+                                      "release:03d}_{version:02d}.h5",
+                        resolved_string=fname,
+                    )
+
+                rgt = temp['rgt']  # Reference Ground Track
+                cycle = temp['cycle']  # Cycle number
+                f_date = str(temp['datetime'].date())  # Datetime
+
+                paras = [rgt, f_date, cycle, bbox_i, self.product]
+                paras_list.append(paras)
+
+        return paras_list
+
+    def request_OA_data(self, paras):
+        """
+        Request data from OpenAltimetry based on API:
+        https://openaltimetry.org/data/swagger-ui/#/
+
+        :param paras: one parameter list for OpenAltimetry API request
+        :return: xarray dataset
+        """
+
+        base_url = 'https://openaltimetry.org/data/api/icesat2/level3a'
+
+        trackId = paras[0]
+        Date = paras[1]
+        cycle = paras[2]
+        bbox = paras[3]
+        product = paras[4]
+
+        # Generate API
+        payload = {'product': product.lower(),
+                   'endDate': Date,
+                   'minx': str(bbox[0]),
+                   'miny': str(bbox[1]),
+                   'maxx': str(bbox[2]),
+                   'maxy': str(bbox[3]),
+                   'trackId': trackId,
+                   'outputFormat': 'json'}  # default return all six beams
+
+        # request OpenAltimetry
+        r = requests.get(base_url, params=payload)
+
+        # get elevation data
+        elevation_data = r.json()
```

Comment:
I feel like there might be a way to simplify data loading from JSON using [`pandas.json_normalize`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.json_normalize.html) (did some work a while ago at https://github.com/weiji14/deepicedrain/blob/v0.4.0/atl06_play.py#L507), but will need to play around a bit to ensure it works. Stay tuned on this!